docs/src/proposals/refactor-banking-buffer.md

```diff
@@ -0,0 +1,253 @@
+Propose to Refactor banking stage buffer
+
+# Motivation
+
+One of the limitations of current MinMaxHeap based banking-stage buffer is that it does not
+include information of account contention, so it is possible that prepared batches are highly
+contended, in extream case, all 64 packets in a batch write to same account.
+
+This potential inefficiency was partly due to deserialized packets are not sanitized into
+transactions until later time, therefore at the time of buffering, banking stage does not have
+transaction's writable accounts information.
+
+Since moving packets' deserialization and sanitization closer to the point of receiving becomes
+necessary, it is now possible to include write accounts info in priority sorted buffer, so
+banking threads can prepare batches of non-conflicting transactions for bank processing.
+
+# Proposal
+
+Buffer can be a structure of: `BTreeMap<Priority, Vec<AccountBuckets>>`, where AccountBuckets is
+a collection of FIFO queues of packets, each FIFO queue contains packets write to same account.
+
+For example: tx1 and tx3 have priority `99`, tx1 writes to A and B, tx3 writes C, would look
+like this: 
+[
+  (99, [ ([tx1, ...]), // txs write to account A
+         ([tx1, ...]), // txs write to account B
+         ([tx3, ...]), // txs write to account C
+         ... ...
+       ]
+  ),
+  (98, [ ([tx1, ...]), // txs write to account A
+         ([tx2, ...]), // txs write to account B
+         ([tx3, ...]), // txs write to account D
+         ... ...
+       ]
+  ),
+  (...)
+]
+
+When consume this buffer, banking thread prepare a batch of non-conflicting transactions from
+the top of priority, pop the oldest packet from non-conflicting account's bucket. In above
+example, a batch could look like `[tx1, tx3, tx2, ...]`
+
+# concerns
+
+1. Too many priority level may hurt performance, 
+2. Multi Iterator applied on top of current MinMaxHeap buffer archives same result; There are
+   few benefits of refactoring buffer from clean code perspective, and perhaps some performance
+   improvement. 
+
+# bench tests
+We have a set of bench tests for `unprocessed_packets_buffer`, as well as `baning-bench`, both
+can be used to perform side-by-side comparison.
+
+# pseudo code
+```
+use std::collections::BTreeMap;
+
+/// Packets
+struct DeserializedPacket {
+    original_packet: Packet,
+    sanitized_transaction: Option<Arc<SanitizedTransaction>>,
+    priority: Priority,
+    writable_accounts: Vec<&Pubkey>,
+    attempted: bool,   // default false, mark to true when selected into batch for processing
+    retryable: bool,   // default false, mark to true when packet is determined as retryable
+    forwarded: bool,
+    // ... ...
+}
+
+/// a FIFO queue for packets that write to same account at same priority
+struct Packets {
+    packets: deque<Arc<DeserializedPacket>>,
```

Comment:
quick note - if we're storing `attempted, retryable, forwarded` on the `DeserializedPacket` we shouldn't wrap `DeserializedPacket` in an `Arc` - need some interior mutability (refcell?).